posts/inside-rust/2023-11-10-parallel-rustc.md

```diff
@@ -0,0 +1,239 @@
+is occuring, intraprocess parallelism will be limited appropriately, and
+the number of threads will not exceed the number of cores.
+
+## How to use it
+
+As of (XXX: date enabled) the nightly compiler is [shipped with the parallel
+front-end enabled](https://github.com/rust-lang/rust/pull/117435). However,
+**by default it runs in single-threaded mode** and won't reduce compile times.
+Keen users who want to try multi-threaded mode should use the `-Z threads=8`
+option.
+
+This default may be surprising. Why parallelize the front-end and then run it
+in single-threaded mode? The answer is simple: caution. This is a big change!
+The parallel front-end has a lot of new code. Single-threaded mode exercises
+most of the new code, but excludes the possibility of threading bugs such as
+deadlocks that still occasionally affect multi-threaded mode. Parallel
+execution is harder to get right than serial execution, even in Rust. For this
+reason the parallel front-end also won't be shipped in Beta or Stable releases
+for some time.
+
+### Performance effects
+
+When the parallel front-end is run in single-threaded mode, compilation times
+are typically within 2% of the serial front-end, which should be barely
+noticeable.
+
+When the parallel front-end is run in multi-threaded mode with `-Z threads=8`,
+out [measurements on real-world
+code](https://github.com/rust-lang/compiler-team/issues/681) on show that
+compile times can be reduced by up to 50%, though the affects vary widely and
+depend greatly on the characteristics of the code being compiled and the build
+configuration. For example, dev builds are likely to see bigger improvements
+than release builds because release builds usually spend more time doing
+optimizations in the back-end. A small number of cases compile more slowly in
+multi-threaded mode than single-threaded mode.
+
+We recommend eight threads because this is the configuration we have tested the
+most and it is known to give good results. Values lower than eight will see
+smaller benefits. Values greater than eight will give diminishing returns and
+may even give worse performance.
+
+If a 50% improvement seems low when going from one to eight threads, recall
+from the explanation above that the front-end only accounts for part of compile
+times, and the back-end is already parallel. You can't beat [Amdahl's
+Law](https://en.wikipedia.org/wiki/Amdahl%27s_law).
+
+Memory usage can increase significantly in multi-threaded mode. This is
+unsurprising given that various tasks, each of which requires a certain amount
+of memory, are now executing in parallel. We have seen increases of up to 35%.
+
+### Correctness
+
+Reliability in single-threaded mode should be high.
+
+In multi-threaded mode there are some known bugs, including deadlocks. If
+compilation hangs, you have probably hit one of them.
+
+### Feedback
+
+If you have any problems with the parallel front-end, please [file an issue
```

Comment:
Currently panics with `no ImplicitCtxt stored in tls` and `deadlock detected` are known (no real deadlocks). Perhaps we should explicitly mention these and encourage reporting if users find other bugs?

Reply:
I see https://github.com/rust-lang/rust/issues/111522 for the `ImplicitCtxt` one. Do we have an open issue for `deadlock detected`?

Reply:
https://github.com/rust-lang/rust/issues/115223 and https://github.com/rust-lang/rust/issues/111521. I renamed the panic in https://github.com/rust-lang/rust/pull/115668. @matthiaskrgr Could you rename the issues so we don't gain extra ones?

Reply:
Hi, sorry, rename what to what?

Reply:
I've tweaked the feedback section to say "If your problem does not match any of the existing issues, please file a new issue." Is that good enough?

Reply:
That's probably fine.